### PR TITLE
FormSettings exception if plugin has no setting

### LINKED ISSUE #7154: [NBug] FormSettings null in PluginRegistry.Plugins =>GetSettings()
## Current behaviour

A colleague had performance problems (a separate issue) and I asked him to test a private debug build done by myself (no changes related to this functionality), running from a folder (portable not set). The change worked OK, but he could not enter settings.

Exception when starting settings in private build
https://github.com/gitextensions/gitextensions/blob/master/GitUI/CommandsDialogs/FormSettings.cs#L115

It seems like some plugin returns null settings.
Obviously with a private build, this will not be kept open for long unless someone has a suggestion. Something like the following could work
var pluginEntries = PluginRegistry.Plugins
                .Where(p => **p.GetSettings() != null &&** p.GetSettings().Any())

## Expected behaviour

No exception

## Steps to reproduce

See above - cannot do it myself

## Error Details
```
System.ArgumentNullException: Value cannot be null.
Parameter name: source
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source)
   at GitUI.CommandsDialogs.FormSettings.<>c.<.ctor>b__12_0(IGitPlugin p) in F:\dev\gc\gitextensions_4\GitUI\CommandsDialogs\FormSettings.cs:line 116
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__1.MoveNext()
   at GitUI.CommandsDialogs.FormSettings..ctor(GitUICommands commands, SettingsPageReference initialPage) in F:\dev\gc\gitextensions_4\GitUI\CommandsDialogs\FormSettings.cs:line 120
   at GitUI.CommandsDialogs.FormSettings.ShowSettingsDialog(GitUICommands uiCommands, IWin32Window owner, SettingsPageReference initialPage) in F:\dev\gc\gitextensions_4\GitUI\CommandsDialogs\FormSettings.cs:line 140
   at GitUI.GitUICommands.<>c__DisplayClass109_0.<StartSettingsDialog>g__Action|0() in F:\dev\gc\gitextensions_4\GitUI\GitUICommands.cs:line 920
   at GitUI.GitUICommands.DoActionOnRepo(IWin32Window owner, Boolean requiresValidWorkingDir, Boolean changesRepo, EventHandler`1 preEvent, EventHandler`1 postEvent, Func`1 action) in F:\dev\gc\gitextensions_4\GitUI\GitUICommands.cs:line 304
   at GitUI.GitUICommands.StartSettingsDialog(IWin32Window owner, SettingsPageReference initialPage) in F:\dev\gc\gitextensions_4\GitUI\GitUICommands.cs:line 925
   at GitUI.CommandsDialogs.FormBrowse.OnShowSettingsClick(Object sender, EventArgs e) in F:\dev\gc\gitextensions_4\GitUI\CommandsDialogs\FormBrowse.cs:line 1441
   at System.Windows.Forms.ToolStripItem.RaiseEvent(Object key, EventArgs e)
   at System.Windows.Forms.ToolStripMenuItem.OnClick(EventArgs e)
   at System.Windows.Forms.ToolStripItem.HandleClick(EventArgs e)
   at System.Windows.Forms.ToolStripItem.HandleMouseUp(MouseEventArgs e)
   at System.Windows.Forms.ToolStrip.OnMouseUp(MouseEventArgs mea)
   at System.Windows.Forms.ToolStripDropDown.OnMouseUp(MouseEventArgs mea)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ToolStrip.WndProc(Message& m)
   at System.Windows.Forms.ToolStripDropDown.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

## Environment
- Git Extensions 3.3.0
- Build 2e1270aff7847a0b942d126e4a44855d4aa9cb81 (Dirty)
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.17763.0
- .NET Framework 4.8.3928.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by gerhardol**
Fixes #7154

## Proposed changes
Null check to avoid exception

## Test methodology 
I do not really have access to the PC that saw this error, but have got the user to test this fix.
While it would be interesting to see why it occurred at all, it is not really feasible.

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by pmiossec** [comment]
We have a real problem in the life cycle of settings controls which is inconsistent. 

And this '.Any()' is part of the problem.

We didn't see it because we do some basic things until now with settings but I hit the wall when developing #7060. 

I tried to fix that in this PR and, even if I'm not 100% satisfied, this commit
https://github.com/gitextensions/gitextensions/commit/783eac318ae2177aac168f5c13dd524f46a712bf should also fix it.

**Comment by gerhardol** [comment]
> We didn't see it because we do some basic things until now with settings but I hit the wall when developing #7060.

That is a better solution for this issue.
I set this to WIP and expect to close it.

**Comment by gerhardol** [comment]
I cherry-picked 783eac3 by @pmiossec - maybe we want that in 3.2.x too (one issue reported except mine).

File: ResourceManager/GitPluginBase.cs [hunk 1]
```diff
@@ -7,6 +7,16 @@
 
 namespace ResourceManager
 {
+    [UsedImplicitly]
+    public abstract class GitPluginWithoutSettingsBase : GitPluginBase
```
Comment (RussKie) [inline ResourceManager/GitPluginBase.cs]: Please move to own file

File: ResourceManager/GitPluginBase.cs [hunk 2]
```diff
@@ -7,6 +7,16 @@
...
+    {
+        public new bool HasSettings => false;
+        public override IEnumerable<ISetting> GetSettings()
```
Comment (RussKie) [inline ResourceManager/GitPluginBase.cs]:
```suggestion

        public override IEnumerable<ISetting> GetSettings()
```

File: ResourceManager/GitPluginBase.cs [hunk 3]
```diff
@@ -7,6 +7,16 @@
 
 namespace ResourceManager
 {
+    [UsedImplicitly]
+    public abstract class GitPluginWithoutSettingsBase : GitPluginBase
+    {
+        public new bool HasSettings => false;
```
Comment (RussKie) [inline ResourceManager/GitPluginBase.cs]: Why are we shadowing the property? We have full control over the code.

It won't work the way you expect if the code is written against the parent type, e.g.
image

In general I suggest avoiding shadowing members unless there is no other way.

File: ResourceManager/GitPluginBase.cs (hunk 1 above)
Comment (RussKie) [inline ResourceManager/GitPluginBase.cs]: Do we _really_ need the new type?
Is there a reason we can't pass a flag to the ctor indicating whether a plugin has settings or not?
```cs
    [Export(typeof(IGitPlugin))]
    public class GitImpactPlugin : GitPluginBase, IGitPluginForRepository
    {
        public GitImpactPlugin(bool hasSettings)
        {
            HasSettings = hasSettings;
        }

        public GitImpactPlugin()
            : this(true)
        {
        }

        public bool HasSettings { get; }
```

**Comment by gerhardol** [comment]
@pmiossec 
You or me? (this is your change, better than my original fix)
I would like to add this to 3.2.x

**Comment by pmiossec** [comment]
@gerhardol done

File: GitUI/CommandsDialogs/SettingsDialog/Plugins/PluginSettingsPage.cs [hunk 4]
```diff
@@ -58,14 +58,14 @@ private IEnumerable<ISetting> GetSettings()
                 throw new ApplicationException();
             }
 
-            return _gitPlugin.GetSettings();
+            return _gitPlugin.HasSettings ? _gitPlugin.GetSettings() : Enumerable.Empty<ISetting>();
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Plugins/PluginSettingsPage.cs]: Will this work?
```suggestion
            return _gitPlugin.HasSettings ? _gitPlugin.GetSettings() : Array.Empty<ISetting>();
```

If I'm not mistaken, in .NET Core these were optimised, in .NET Framework `Enumerable.Empty<T>()` is less efficient than `Array.Empty<T>`

Comment (pmiossec) [reply GitUI/CommandsDialogs/SettingsDialog/Plugins/PluginSettingsPage.cs]: yes. done.

**Comment by RussKie** [comment]
Thank you gents 👍

**Pull request merged**